File: src/application/ports/ExchangeRateProvider.ts
import { CurrencyPair } from '../../domain/CurrencyPair.js';

export interface ExchangeRateQuote {
  pair: CurrencyPair;
  rate: number;
}

export interface ExchangeRateProvider {
  getQuote(pair: CurrencyPair): Promise<ExchangeRateQuote | null>;
}

File: internal/application/ports/exchange_rate_provider.go
package ports

import (
	"context"

	"github.com/shopspring/decimal"

	"github.com/example/payment-gateway/internal/domain"
)

// ExchangeRateProvider resolves exchange rates for supported currency pairs.
type ExchangeRateProvider interface {
	GetRate(ctx context.Context, pair domain.CurrencyPair) (decimal.Decimal, error)
}
